File: Medium/Implement Trie (Prefix Tree).cpp
/*
Implement a trie with insert, search, and startsWith methods.

Example:

Trie trie = new Trie();

trie.insert("apple");
trie.search("apple");   // returns true
trie.search("app");     // returns false
trie.startsWith("app"); // returns true
trie.insert("app");   
trie.search("app");     // returns true

Note:
    You may assume that all inputs are consist of lowercase letters a-z.
    All inputs are guaranteed to be non-empty strings.
*/

class Trie {
private:
    struct TrieNode{
        struct TrieNode *child[26];
        bool isEnd;
    };
    
    TrieNode *searchPrefix(string word) {
        TrieNode *node = root;
        for(char ch : word) {
            if(!node->child[ch - 'a']){
                return NULL;
            }
            
            node = node->child[ch - 'a'];
        }
        
        return node;
    }
public:
    /** Initialize your data structure here. */
    TrieNode *root = NULL;
    
    Trie() {
        root = new TrieNode();   
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode *node = root;
        
        for(char ch : word) {
            if(!node->child[ch - 'a']) {
                node->child[ch - 'a'] = new TrieNode();
            }
            
            node = node->child[ch - 'a'];
        }
    
        node->isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode *node = searchPrefix(word);
        
        return (node != NULL && node->isEnd) ;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        return (searchPrefix(prefix) != NULL);
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: Medium/Implement Trie (Prefix Tree).java
/*
A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. 
There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

Trie() Initializes the trie object.
void insert(String word) Inserts the string word into the trie.
boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), 
and false otherwise.
boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.
 

Example 1:

Input
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
Output
[null, null, true, false, true, null, true]

Explanation
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // return True
trie.search("app");     // return False
trie.startsWith("app"); // return True
trie.insert("app");
trie.search("app");     // return True
 

Constraints:

1 <= word.length, prefix.length <= 2000
word and prefix consist only of lowercase English letters.
At most 3 * 10^4 calls in total will be made to insert, search, and startsWith.
*/

class Trie {

    private TrieNode root;
    public Trie() {
        this.root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = this.root;
        
        for (char ch : word.toCharArray()) {
            if (!node.containsKey(ch)) {
                node.put(ch, new TrieNode());
            }
            
            node = node.get(ch);
        }
        
        node.setEnd();
    }
    
    public boolean search(String word) {
        TrieNode node = searchPrefix(word);
        
        return Objects.nonNull(node) && node.isEnd();
        
    }
    
    public boolean startsWith(String prefix) {
        TrieNode node = searchPrefix(prefix);
        
        return Objects.nonNull(node);
    }
    
    private TrieNode searchPrefix(String word) {
        TrieNode node = this.root;
        
        for (char ch : word.toCharArray()) {
            if (node.containsKey(ch)) {
                node = node.get(ch);
            } else {
                return null;
            }
        }
        
        return node;
    }
}

class TrieNode {
    private int limit = 26;
    private TrieNode[] links;
    private boolean isEnd;
    
    public TrieNode() {
        this.links = new TrieNode[this.limit];  
    }
    
    public boolean containsKey(char key) {
        return Objects.nonNull(this.links[key - 'a']);
    }
    
    public TrieNode get(char key) {
        return this.links[key - 'a'];
    }
    
    public void put(char key, TrieNode node) {
        this.links[key - 'a'] = node;
    }
    
     
    public void setEnd() {
        isEnd = true;
    }
    
    public boolean isEnd(){
        return isEnd;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */